File: implementations/graphql-java-kickstart/src/main/java/graphql/kickstart/federation/compatibility/model/Product.java
package graphql.kickstart.federation.compatibility.model;

import static graphql.kickstart.federation.compatibility.model.ProductResearch.FEDERATION_STUDY;
import static graphql.kickstart.federation.compatibility.model.ProductResearch.STUDIO_STUDY;
import static graphql.kickstart.federation.compatibility.model.User.DEFAULT_USER;

import java.util.ArrayList;
import java.util.List;
import org.jetbrains.annotations.NotNull;

public record Product(String id, String sku, String pkg, ProductVariation variation, ProductDimension dimensions, User createdBy, String notes, List<ProductResearch> research) {

    public static List<Product> PRODUCTS = List.of(
            new Product("apollo-federation", "federation", "@apollo/federation", "OSS", FEDERATION_STUDY),
            new Product("apollo-studio", "studio", "", "platform", STUDIO_STUDY)
    );

    public Product(String id, String sku, String pkg, String variationId, ProductResearch research) {
        this(id, sku, pkg, new ProductVariation(variationId), new ProductDimension("small", 1, "kg"), DEFAULT_USER, "", List.of(research));
    }

    public String getPackage() {
        return pkg;
    }

    public static Product resolveById(@NotNull String productId) {
        for (Product product : PRODUCTS) {
            if (product.id().equals(productId)) {
                return product;
            }
        }
        return null;
    }
}

File: implementations/sangria/src/main/scala/model/Product.scala
package model

import io.circe.Decoder
import io.circe.generic.semiauto.deriveDecoder

case class ProductVariation(id: ID)
object ProductVariation {
  implicit val jsonDecoder: Decoder[ProductVariation] = deriveDecoder[ProductVariation]
}

case class ProductDimension(size: Option[String], weight: Option[Double], unit: Option[String])

case class CaseStudy(
    caseNumber: ID,
    description: Option[String]
)

case class ProductResearch(
    study: CaseStudy,
    outcome: Option[String] = None
)

case class Product(
    id: ID,
    sku: Option[String],
    `package`: Option[String],
    variation: Option[ProductVariation],
    dimensions: Option[ProductDimension],
    createdBy: Option[User],
    research: List[ProductResearch],
    notes: Option[String] = None
)

case class DeprecatedProduct(
    sku: String,
    `package`: String,
    reason: Option[String],
    createdBy: Option[User]
)
